drv/ArrayList.drv

```diff
@@ -450,11 +450,9 @@ public class ARRAY_LIST KEY_GENERIC extends ABSTRACT_LIST KEY_GENERIC implements
 #else
 		if (wrapped) a = ARRAYS.ensureCapacity(a, capacity, size);
 		else {
-			if (capacity > a.length) {
-				final Object[] t = new Object[capacity];
-				System.arraycopy(a, 0, t, 0, size);
-				a = (KEY_GENERIC_TYPE[])t;
-			}
+			final Object[] t = new Object[capacity];
+			System.arraycopy(a, 0, t, 0, size);
```

Comment:
Shouldn't Arrays.copyOf be applicable here? This can potentially be faster by a) calling realloc and b) avoid initializing the new array with nulls.

Reply:
`Arrays.copyOf` copies the entire array data rather than copying an array data of the `size` of the list, which may result in performance differences. In the case of copying the entire array data, their time consumption is the same:
```
@Fork(1)
@State(Scope.Thread)
@Warmup(iterations = 2)
@Measurement(iterations = 2)
@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
public class ArrayCopyBenchmark {
    private int oldCapacity = 1000;
    private int newCapacity = oldCapacity + oldCapacity >> 1;

    private int[] a;

    @Setup
    public void setup() {
        a = new int[oldCapacity];
        Arrays.setAll(a, IntUnaryOperator.identity());
    }


    @Benchmark
    public void grow() {
        int[] t = new int[newCapacity];
        System.arraycopy(a, 0, t, 0, oldCapacity);
        a = t;
    }

    @Benchmark
    public void copyOf() {
        a = Arrays.copyOf(a, newCapacity);
    }
}
```
Output:
```
# JMH version: 1.37
# VM version: JDK 17.0.10, Java HotSpot(TM) 64-Bit Server VM, 17.0.10+11-LTS-240
# VM invoker: C:\Program Files\jdk-17\bin\java.exe
# VM options: -javaagent:C:\Program Files\IntelliJ IDEA 2023.2.5\lib\idea_rt.jar=4477:C:\Program Files\IntelliJ IDEA 2023.2.5\bin -Dfile.encoding=UTF-8
# Blackhole mode: compiler (auto-detected, use -Djmh.blackhole.autoDetect=false to disable)
# Warmup: 2 iterations, 10 s each
# Measurement: 2 iterations, 10 s each
# Timeout: 10 min per iteration
# Threads: 1 thread, will synchronize iterations
# Benchmark mode: Average time, time/op

Benchmark                  Mode  Cnt    Score   Error  Units
ArrayCopyBenchmark.copyOf  avgt    2  239.428          ns/op
ArrayCopyBenchmark.grow    avgt    2  241.533          ns/op
```

Reply:
> Arrays.copyOf copies the entire array data rather than copying an array data of the size of the list, which may result in performance differences.

note that `new Object[capacity]` needs to fill the array with `null`

But yes, my comment was under the assumption that `size == a.length` which it need not be. I guess it wouldn't make a difference.

Reply:
You are right. In cases where `size != a.length`, `Arrays.copyOf` is better.
```
@Fork(1)
@State(Scope.Thread)
@Warmup(iterations = 3)
@Measurement(iterations = 10)
@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
public class ArrayCopyBenchmark {
    private int size = 50000;
    private int oldCapacity = 100000;
    private int newCapacity = oldCapacity + oldCapacity >> 1;

    private int[] a;

    @Setup
    public void setup() {
        a = new int[oldCapacity];
        Arrays.setAll(a, IntUnaryOperator.identity());
    }


    @Benchmark
    public int[] ensureCapacity() {
        int[] t = new int[newCapacity];
        System.arraycopy(a, 0, t, 0, size);
        a = t;
        return a;
    }

    @Benchmark
    public int[] copyOf() {
        a = Arrays.copyOf(a, newCapacity);
        return a;
    }
}
```
Output:
```
# JMH version: 1.37
# VM version: JDK 17.0.10, Java HotSpot(TM) 64-Bit Server VM, 17.0.10+11-LTS-240
# VM invoker: C:\Program Files\jdk-17\bin\java.exe
# VM options: -javaagent:C:\Program Files\IntelliJ IDEA 2023.2.5\lib\idea_rt.jar=6866:C:\Program Files\IntelliJ IDEA 2023.2.5\bin -Dfile.encoding=UTF-8
# Blackhole mode: compiler (auto-detected, use -Djmh.blackhole.autoDetect=false to disable)
# Warmup: 3 iterations, 10 s each
# Measurement: 10 iterations, 10 s each
# Timeout: 10 min per iteration
# Threads: 1 thread, will synchronize iterations
# Benchmark mode: Average time, time/op

Benchmark                          Mode  Cnt      Score    Error  Units
ArrayCopyBenchmark.copyOf          avgt   10  26417.144 ± 83.420  ns/op
ArrayCopyBenchmark.ensureCapacity  avgt   10  25251.914 ± 65.468  ns/op
```